REGRESSION(265569@main): All path segments should be coded by StreamConnectionEncoder

shallawa opened the pull request:
#### 7dc4e7c7d38b1bc12aa3b7ba168e23699bc5fe6e
<pre>
REGRESSION(265569@main): All path segments should be coded by StreamConnectionEncoder
<a href="https://bugs.webkit.org/show_bug.cgi?id=258734">https://bugs.webkit.org/show_bug.cgi?id=258734</a>
rdar://111570644

Reviewed by Wenson Hsieh.

Detemplatize the composite path segments. Add StreamConnectionEncoder as additional
encoder for these structures. This should recover part of the CanvasLines regression
which was caused by 265569@main.

* Source/WebCore/platform/graphics/PathSegmentData.cpp:
(WebCore::calculateArcToEndPoint):
(WebCore::PathArcTo::calculateEndPoint const):
(WebCore::PathArcTo::extendBoundingRect const):
(WebCore::PathDataLine::calculateEndPoint const):
(WebCore::PathDataLine::extendFastBoundingRect const):
(WebCore::PathDataLine::extendBoundingRect const):
(WebCore::PathDataLine::addToImpl const):
(WebCore::PathDataLine::applyElements const):
(WebCore::operator&lt;&lt;):
(WebCore::PathDataQuadCurve::calculateEndPoint const):
(WebCore::PathDataQuadCurve::extendFastBoundingRect const):
(WebCore::PathDataQuadCurve::extendBoundingRect const):
(WebCore::PathDataQuadCurve::addToImpl const):
(WebCore::PathDataQuadCurve::applyElements const):
(WebCore::PathDataBezierCurve::calculateEndPoint const):
(WebCore::PathDataBezierCurve::extendFastBoundingRect const):
(WebCore::PathDataBezierCurve::extendBoundingRect const):
(WebCore::PathDataBezierCurve::addToImpl const):
(WebCore::PathDataBezierCurve::applyElements const):
(WebCore::PathDataArc::calculateEndPoint const):
(WebCore::PathDataArc::extendFastBoundingRect const):
(WebCore::PathDataArc::extendBoundingRect const):
(WebCore::PathDataArc::addToImpl const):
(WebCore::DataType2&gt;::calculateEndPoint const): Deleted.
(WebCore::DataType2&gt;::extendFastBoundingRect const): Deleted.
(WebCore::DataType2&gt;::extendBoundingRect const): Deleted.
(WebCore::DataType2&gt;::addToImpl const): Deleted.
(WebCore::DataType2&gt;::applyElements const): Deleted.
* Source/WebCore/platform/graphics/PathSegmentData.h:
(WebCore::PathDataArc::applyElements const):
(WebCore::operator&lt;&lt;): Deleted.
* Source/WebCore/platform/graphics/PathStream.cpp:
(WebCore::PathStream::lastIfMoveTo const):
(WebCore::PathStream::addLineTo):
(WebCore::PathStream::addQuadCurveTo):
(WebCore::PathStream::addBezierCurveTo):
(WebCore::PathStream::addArcTo):
(WebCore::PathStream::mergeIntoComposite): Deleted.
* Source/WebCore/platform/graphics/PathStream.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::strokePath):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
(WebCore::DisplayList::StrokeLine::StrokeLine):
* Source/WebKit/GPUProcess/graphics/PathSegment.serialization.in:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
(): Deleted.
* Tools/TestWebKitAPI/Tests/WebCore/DisplayListTests.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/265680@main">https://commits.webkit.org/265680@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c0770be012af738eb5f3620e3e90e9d009dd0774

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11572 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11776 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12137 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13218 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11032 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11593 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14162 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11756 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13901 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11736 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12593 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9810 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13639 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9885 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10505 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17645 "Passed tests") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10958 "1 api test failed or timed out") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10659 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13839 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11057 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9112 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10234 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2794 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14511 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10914 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/platform/graphics/PathSegmentData.cpp:
```diff
@@ -532,43 +532,159 @@ WTF::TextStream& operator<<(WTF::TextStream& ts, const PathRoundedRect& data)
+
+void PathDataLine::extendBoundingRect(const FloatPoint&, const FloatPoint&, FloatRect& boundingRect) const
+{
+    boundingRect.extend(start);
+    boundingRect.extend(end);
```
Ditto.

whsieh commented on Source/WebCore/platform/graphics/PathSegmentData.cpp:
```diff
@@ -532,43 +532,159 @@ WTF::TextStream& operator<<(WTF::TextStream& ts, const PathRoundedRect& data)
+
+void PathDataQuadCurve::extendFastBoundingRect(const FloatPoint&, const FloatPoint&, FloatRect& boundingRect) const
+{
+    boundingRect.extend(start);
+    boundingRect.extend(controlPoint);
+    boundingRect.extend(endPoint);
```
Ditto.

whsieh commented on Source/WebCore/platform/graphics/PathSegmentData.cpp:
```diff
@@ -532,43 +532,159 @@ WTF::TextStream& operator<<(WTF::TextStream& ts, const PathRoundedRect& data)
+void PathDataQuadCurve::extendBoundingRect(const FloatPoint&, const FloatPoint&, FloatRect& boundingRect) const
+{
+    auto extremity = calculateQuadraticExtremity(start, controlPoint, endPoint);
+    boundingRect.extend(start);
+    boundingRect.extend(extremity);
+    boundingRect.extend(endPoint);
```
Ditto, re: my other comments — perhaps we can just put this behind a single helper method, or even a new method on `WebCore::FloatRect`? Something like:

```
template<typename... PointType>
void extendBounds(FloatRect& bounds, const PointType&... points)
{
    float minX = std::min({ bounds.x(), points.x()... });
    float minY = std::min({ bounds.y(), points.y()... });
    float maxX = std::max({ bounds.maxX(), points.x()... });
    float maxY = std::max({ bounds.maxY(), points.y()... });
    bounds.setLocation({ minX, minY });
    bounds.setSize({ maxX - minX, maxY - minY });
}
```

...so you can simply do this:

```
extendBounds(boundingRect, currentPoint, controlPoint1, calculateArcToEndPoint(currentPoint, controlPoint1, controlPoint2, radius));
```

or alternately:

```
boundingRect.extend(currentPoint, controlPoint1, calculateArcToEndPoint(currentPoint, controlPoint1, controlPoint2, radius));
```

whsieh commented on Source/WebCore/platform/graphics/PathSegmentData.cpp:
```diff
@@ -310,7 +310,7 @@ void PathArcTo::extendBoundingRect(const FloatPoint& currentPoint, const FloatPo
 {
     boundingRect.extend(currentPoint);
     boundingRect.extend(controlPoint1);
-    boundingRect.extend(calculateEndPoint(currentPoint));
+    boundingRect.extend(calculateArcToEndPoint(currentPoint, controlPoint1, controlPoint2, radius));
```
I wonder if we could make this _ever so slightly_ more efficient by rewriting this to make it only call `maxX`/`maxY` and `setLocationAndSizeFromEdges` once.

whsieh commented on Source/WebCore/platform/graphics/PathSegmentData.cpp:
```diff
@@ -532,43 +532,159 @@ WTF::TextStream& operator<<(WTF::TextStream& ts, const PathRoundedRect& data)
+
+void PathDataArc::extendFastBoundingRect(const FloatPoint&, const FloatPoint&, FloatRect& boundingRect) const
+{
+    boundingRect.extend(start);
+    boundingRect.extend(controlPoint1);
+    boundingRect.extend(controlPoint2);
```
(Could use the new helper method or function here as well)

whsieh commented on Source/WebCore/platform/graphics/PathSegmentData.cpp:
```diff
@@ -532,43 +532,159 @@ WTF::TextStream& operator<<(WTF::TextStream& ts, const PathRoundedRect& data)
+
+void PathDataArc::extendBoundingRect(const FloatPoint&, const FloatPoint&, FloatRect& boundingRect) const
+{
+    boundingRect.extend(start);
+    boundingRect.extend(controlPoint1);
+    boundingRect.extend(calculateArcToEndPoint(start, controlPoint1, controlPoint2, radius));
```
Ditto.

whsieh commented on Source/WebCore/platform/graphics/PathStream.cpp:
```diff
@@ -102,34 +96,34 @@ void PathStream::moveTo(const FloatPoint& point)
+    if (const auto* moveTo = lastIfMoveTo())
+        m_segments.last() = { PathDataLine { moveTo->point, point } };
+    else
+        m_segments.append(PathLineTo { point });
```
Nit - you _might_ be able to avoid some code duplication here and in the methods below, by introducing a templated helper function.

shallawa committed:
REGRESSION(265569@main): All path segments should be coded by StreamConnectionEncoder
https://bugs.webkit.org/show_bug.cgi?id=258734
rdar://111570644

Reviewed by Wenson Hsieh.

Detemplatize the composite path segments. Add StreamConnectionEncoder as additional
encoder for these structures. This should recover part of the CanvasLines regression
which was caused by 265569@main.

* Source/WebCore/platform/graphics/PathSegmentData.cpp:
(WebCore::calculateArcToEndPoint):
(WebCore::PathArcTo::calculateEndPoint const):
(WebCore::PathArcTo::extendBoundingRect const):
(WebCore::PathDataLine::calculateEndPoint const):
(WebCore::PathDataLine::extendFastBoundingRect const):
(WebCore::PathDataLine::extendBoundingRect const):
(WebCore::PathDataLine::addToImpl const):
(WebCore::PathDataLine::applyElements const):
(WebCore::operator<<):
(WebCore::PathDataQuadCurve::calculateEndPoint const):
(WebCore::PathDataQuadCurve::extendFastBoundingRect const):
(WebCore::PathDataQuadCurve::extendBoundingRect const):
(WebCore::PathDataQuadCurve::addToImpl const):
(WebCore::PathDataQuadCurve::applyElements const):
(WebCore::PathDataBezierCurve::calculateEndPoint const):
(WebCore::PathDataBezierCurve::extendFastBoundingRect const):
(WebCore::PathDataBezierCurve::extendBoundingRect const):
(WebCore::PathDataBezierCurve::addToImpl const):
(WebCore::PathDataBezierCurve::applyElements const):
(WebCore::PathDataArc::calculateEndPoint const):
(WebCore::PathDataArc::extendFastBoundingRect const):
(WebCore::PathDataArc::extendBoundingRect const):
(WebCore::PathDataArc::addToImpl const):
(WebCore::DataType2>::calculateEndPoint const): Deleted.
(WebCore::DataType2>::extendFastBoundingRect const): Deleted.
(WebCore::DataType2>::extendBoundingRect const): Deleted.
(WebCore::DataType2>::addToImpl const): Deleted.
(WebCore::DataType2>::applyElements const): Deleted.
* Source/WebCore/platform/graphics/PathSegmentData.h:
(WebCore::PathDataArc::applyElements const):
(WebCore::operator<<): Deleted.
* Source/WebCore/platform/graphics/PathStream.cpp:
(WebCore::PathStream::lastIfMoveTo const):
(WebCore::PathStream::addLineTo):
(WebCore::PathStream::addQuadCurveTo):
(WebCore::PathStream::addBezierCurveTo):
(WebCore::PathStream::addArcTo):
(WebCore::PathStream::mergeIntoComposite): Deleted.
* Source/WebCore/platform/graphics/PathStream.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::strokePath):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
(WebCore::DisplayList::StrokeLine::StrokeLine):
* Source/WebKit/GPUProcess/graphics/PathSegment.serialization.in:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
(): Deleted.
* Tools/TestWebKitAPI/Tests/WebCore/DisplayListTests.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/265680@main
